taichi/ir/ir.h

```diff
@@ -537,6 +554,36 @@ class StmtFieldNumeric final : public StmtField {
   }
 };
 
+class StmtFieldSNode final : public StmtField {
+ private:
+  SNode *const &snode;
+
+ public:
+  explicit StmtFieldSNode(SNode *const &snode): snode(snode) {
+  }
+
+  static int get_snode_id(SNode *snode) {
+    if (snode == nullptr)
+      return -1;
+    return snode->id;
+  }
+
+  bool equal(const StmtField *other_generic) const override {
+    if (auto other = dynamic_cast<const StmtFieldSNode *>(other_generic)) {
+      return get_snode_id(snode) == get_snode_id(other->snode);
+    } else {
+      // Different types
+      return false;
+    }
+  }
+};
+
+template <class T, template <class...> class Template>
+struct is_specialization : std::false_type {};
+
+template <template <class...> class Template, class... Args>
+struct is_specialization<Template<Args...>, Template> : std::true_type {};
```

Comment:
Please move these to https://github.com/taichi-dev/taichi/blob/master/taichi/common/meta.h so that other people can reuse :-)